server/app/services/applicant/ApplicantService.java

```diff
@@ -283,6 +290,181 @@ private CompletionStage<ReadOnlyApplicantProgramService> stageAndUpdateIfValid(
+                serviceAreaInclusionGroupFutures =
+                    filteredServiceAreaOptions.stream()
+                        .map(
+                            (option) -> {
+                              return esriClient
+                                  .getServiceAreaInclusionGroup(option, addressLocation)
+                                  .thenApplyAsync(
+                                      (serviceAreaInclusionGroup) -> serviceAreaInclusionGroup,
+                                      httpExecutionContext.current());
+                            })
+                        .collect(ImmutableList.toImmutableList());
+
+            return CompletableFuture.allOf(
+                    serviceAreaInclusionGroupFutures.toArray(
+                        new CompletableFuture[serviceAreaInclusionGroupFutures.size()]))
+                .thenComposeAsync(
+                    (u) -> {
+                      ImmutableList<ServiceAreaInclusion> serviceAreaInclusionGroup =
+                          serviceAreaInclusionGroupFutures.stream()
+                              .map((future) -> future.toCompletableFuture().join())
+                              .flatMap(Collection::stream)
+                              .collect(ImmutableList.toImmutableList());
+                      ImmutableList<ServiceAreaInclusion> newServiceAreaInclusionGroup;
+
+                      if (existingServiceAreaInclusionGroup.size() > 0) {
+                        newServiceAreaInclusionGroup =
+                            Stream.of(existingServiceAreaInclusionGroup, serviceAreaInclusionGroup)
+                                .flatMap(List::stream)
+                                .collect(
+                                    Collectors.toMap(
+                                        ServiceAreaInclusion::getServiceAreaId,
+                                        area -> area,
+                                        (ServiceAreaInclusion a, ServiceAreaInclusion b) ->
+                                            b == null ? a : b))
+                                .values()
+                                .stream()
+                                .collect(ImmutableList.toImmutableList());
+                      } else {
+                        newServiceAreaInclusionGroup = serviceAreaInclusionGroup;
+                      }
+
+                      UpdateServiceArea updateServiceArea =
+                          UpdateServiceArea.create(serviceAreaPath, newServiceAreaInclusionGroup);
+
+                      return stageAndUpdateIfValid(
+                          applicant,
+                          baseUrl,
+                          blockBeforeUpdate,
+                          programDefinition,
+                          updates,
+                          updateServiceArea);
+                    },
+                    httpExecutionContext.current());
+          }
+        }
+      }
+    }
```

Comment:
**suggestion (blocking):** I'd really like to find a way to factor this nearly 100 line if statement into a method so someone who is reading through how updates happen doesn't get caught up in the details of service area checks

Reply:
You could factor the service area logic into a separate class (or method in here, though class is preferable including the service area related private methods below) with a public interface that returns completion stage:

```java
public CompletionStage<Optional<UpdateServiceArea>> getServiceAreaUpdate(...) {
  // do the stuff inside this if statement or
  // return an `Optional.empty()` or throw an exception as needed
}
```

Then this method gets a lot easier to read through:

```java
    if (addressServiceAreaValidationEnabled
        && blockBeforeUpdate.getLeafAddressNodeServiceAreaIds().isPresent()) {
        return serviceAreaUpdateResolver.getServiceAreaUpdate(...)
          .thenApplyAsync((serviceAreaUpdate) -> {
            return stageAndUpdateIfValid(
        applicant, baseUrl, blockBeforeUpdate, programDefinition, updates, updateServiceArea);
          });
   }

  return stageAndUpdateIfValid(
        applicant, baseUrl, blockBeforeUpdate, programDefinition, updates, /* serviceAreaUpdate= */ Optiona.empty());
          });
```

Reply:
Thanks for the suggestion! This really helped clean up the code and readability.